src/bridge/cockpitpipechannel.c

```diff
@@ -483,3 +510,46 @@ cockpit_pipe_channel_open (CockpitTransport *transport,
   json_object_unref (options);
   return channel;
 }
+
+static void
+internal_fd_free (gpointer data)
+{
+  gint fd = GPOINTER_TO_INT (data);
+
+  close (fd);
+}
+
+const gchar *
+cockpit_pipe_channel_add_internal_fd (gint fd)
+{
+  /* We are not multi-threaded. Also don't make this look like normal fd numbers */
+  static guint64 unique = 911111;
+
+  gchar *id;
+
+  if (!internal_fds)
+    internal_fds = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, internal_fd_free);
+
+  id = g_strdup_printf ("file-descriptor-%" G_GUINT64_FORMAT, unique++);
+  g_hash_table_replace (internal_fds, id, GINT_TO_POINTER (fd));
```

Comment:
I'd actually do the assertion here -- as we've discussed above, we do want the IDs  to be unique even if the fds get reused, so this prevents falling into a trap in the future when someone "simplifies" the ID housekeeping.